onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -211,58 +233,37 @@ struct ProviderHostImpl : ProviderHost {
     DataTypeImpl_GetTensorType_float = &DataTypeImpl::GetTensorType<float>;
   }
 
-  std::unique_ptr<Provider_OrtMemoryInfo> OrtMemoryInfo_Create(const char* name_, OrtAllocatorType type_, Provider_OrtDevice* device_, int id_, OrtMemType mem_type_) override {
-    return onnxruntime::make_unique<Provider_OrtMemoryInfo_Impl>(name_, type_, device_ ? static_cast<Provider_OrtDevice_Impl*>(device_)->v_ : OrtDevice(), id_, mem_type_);
-  }
-
   Provider_AllocatorPtr CreateAllocator(const Provider_AllocatorCreationInfo& info) override {
     AllocatorCreationInfo info_real{
-        [&info](int value) {
-          return std::move(static_cast<Provider_IDeviceAllocator_Impl*>(&*info.factory(value))->p_);
+        [&info](int value) -> std::unique_ptr<IDeviceAllocator> {
+          auto allocator = info.factory(value);
+          // If the allocator is a provider interface, we need to wrap it with ProviderAllocator to turn it into an IDeviceAllocator
+          // Otherwise it's really a Provider_IDeviceAllocator_Impl, so we can just unwrap it to get back to the IDeviceAllocator inside
+          if (allocator->IsProviderInterface())
+            return onnxruntime::make_unique<ProviderAllocator>(std::move(allocator));
+          else
+            return std::move(static_cast<Provider_IDeviceAllocator_Impl*>(&*allocator)->p_);
```

Comment:
>std::move [](start = 19, length = 9)

Avoid move on return, let the compiler do the job. [https://stackoverflow.com/questions/11817873/using-stdmove-when-returning-a-value-from-a-function-to-avoid-to-copy](https://stackoverflow.com/questions/11817873/using-stdmove-when-returning-a-value-from-a-function-to-avoid-to-copy)

Reply:
But this wasn't a local variable, so I thought it was better in this case. I can change it as it's just a shared_ptr, but I wanted to bypass one addref/release